Groovy parser fails on array instantiation (e.g. new String[2])

## What version of OpenRewrite are you using?
Current main @ https://github.com/openrewrite/rewrite/commit/9abac8d9b5bd5bd4953569290b6591e91ed130cf



## What is the smallest, simplest way to reproduce the problem?
```java
    @Test
    void variableDeclarationWithArrayInstantiation() {
        rewriteRun(
          groovy(
            """
            def arr = new String[2]
            """
          )
        );
    }
```

I also noticed similar failure with:
```
        rewriteRun(
          groovy(
            """
            def arr = null
            arr ?= new String[2]
            """
          )
        );
```

## What is the full stack trace of any errors you encountered?
```
Caused by: java.lang.StringIndexOutOfBoundsException: Index 23 out of bounds for length 23
	at java.base/jdk.internal.util.Preconditions$1.apply(Preconditions.java:55)
	at java.base/jdk.internal.util.Preconditions$1.apply(Preconditions.java:52)
	at java.base/jdk.internal.util.Preconditions$4.apply(Preconditions.java:213)
	at java.base/jdk.internal.util.Preconditions$4.apply(Preconditions.java:210)
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:98)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:106)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:302)
	at java.base/java.lang.String.checkIndex(String.java:4832)
	at java.base/java.lang.StringLatin1.charAt(StringLatin1.java:46)
	at java.base/java.lang.String.charAt(String.java:1555)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.buildNewArrayDimensions(GroovyParserVisitor.java:725)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitArrayExpression(GroovyParserVisitor.java:689)
	at org.codehaus.groovy.ast.expr.ArrayExpression.visit(ArrayExpression.java:102)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visit(GroovyParserVisitor.java:620)
```

## Are you interested in [contributing a fix to OpenRewrite]
If no one picks it up, I may look into it once have time.

--- a/rewrite-groovy/src/main/java/org/openrewrite/groovy/GroovyParserVisitor.java
+++ b/rewrite-groovy/src/main/java/org/openrewrite/groovy/GroovyParserVisitor.java
@@ -722,7 +722,7 @@ public class GroovyParserVisitor {
             }
             while (true) {
                 int beginBracket = indexOfNextNonWhitespace(cursor, source);
-                if (source.charAt(beginBracket) != '[') {
+                if (beginBracket >= source.length() || source.charAt(beginBracket) != '[') {
                     break;
                 }
 

--- a/rewrite-groovy/src/test/java/org/openrewrite/groovy/GroovyParserTest.java
+++ b/rewrite-groovy/src/test/java/org/openrewrite/groovy/GroovyParserTest.java
@@ -117,4 +117,15 @@ class GroovyParserTest implements RewriteTest {
           )
         );
     }
+
+    @Test
+    void variableDeclarationWithArrayInstantiationAsTheLastStatement() {
+        rewriteRun(
+          groovy(
+            """
+            def arr = new String[2]
+            """
+          )
+        );
+    }
 }

